File: python/53.py
class Solution:
    def maxSubArray(self, nums):
        n = len(nums)  
        if n == 0:
            return 0
        # 这里可以直接使用nums，为了方便我更好的可读使用了dp
        # dp：已当前i为末尾的序列最大序列和
        # 转移方程：dp[i] = max(nums[i], nums[i] + dp[i-1])
        # 解释，以i为结尾有两种情况 [nums[i]], [...nums[i-1], nums[i]]
        dp = [0] * n  
        max_sum = nums[0]
        dp[0] = nums[0]
        for i in range(1, n):
            dp[i] = max(nums[i], nums[i] + dp[i-1])
            max_sum = max(dp[i], max_sum)

        return max_sum

s = Solution()
s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4])

File: cpp/53.cpp
#include <vector>
#include <cmath>
#include <cstdio>
using std::vector;
using std::printf;

#define MAX(a, b) (a) > (b) ? (a) : (b)

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        /*
         * 4 ms	
         * 9.4 MB
         */
        int n = nums.size();
        if (n == 0) {
            return 0;
        }
        /*
         * dp数组的含义：到i为结尾最大子序列和
         * 转移方程： dp[i] = MAX(nums[i], nums[i] + dp[i - 1]);
         * 解释，以i为结尾有两种情况 [nums[i]], [...nums[i-1], nums[i]]
         */
        vector<int> dp(n);  // 这个其实可以直接使用nums代替，这里使用dp是为了更好理解
        dp[0] = nums[0];
        int max_sum = nums[0];
        for (int i = 1; i < n; i++) {
            dp[i] = MAX(nums[i], nums[i] + dp[i - 1]);
            max_sum = MAX(dp[i], max_sum); 
        }
        return max_sum;
    }
};

int main()
{
    Solution s;
    vector<int> vec = {-2, -1};
    printf("%d", s.maxSubArray(vec));
}